rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,223 @@
+
+### Check: detailed features of GOV.UK work
+
+A previous draft of this RFC would have required a detailed audit of the functionality of each app, to ensure we had Smoke or sandboxed E2E tests for each of its features, for some definition of "feature".
+
+This approach was rejected because:
+
+- There is no clear way to demarcate features of an app, let alone GOV.UK as a whole. Given the sheer volume of functionality in our apps, it's not practical to audit or have realistic tests for all of it.
+
+- While we have existing tests for specific features of GOV.UK [[1](https://github.com/alphagov/publishing-e2e-tests/tree/b4a7961cdcd200623a694baf833724a41c98d051/spec/collections_publisher)] [[2](https://github.com/alphagov/smokey/blob/a60df2cde5c886fa6d54e8cd820d90facca89e8e/features/csv_preview.feature)], there is no thorough, positive logic to justify their existence or extension [[1](https://github.com/alphagov/govuk-rfcs/blob/master/rfc-050-do-end-to-end-testing-of-gov-uk-applications.md#solution)] [[2](https://github.com/alphagov/publishing-e2e-tests/blob/f12ce7a9626b38728758cf157ea883907bd89de2/docs/what-belongs-in-these-tests.md)]. This is a weak foundation to try and build upon.
+
+## Consequences
+
+### Being more cautious
+
+We will no longer have the safety blanket of our changes languishing in Integration or Staging. It may take some time for us to get used to this new way of working e.g. by gradually doing more [branch deployments](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#review-your-own-changes).
+
+> **Quote from the trial of CD**: "We'll have to up our game to check things before deploying them. There are always corner cases you don't expect. We're perfectly capable of checking all our changes."
+
+### Automatic deployments
+
+We will enable automatic deployments for the apps listed [here](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/hieradata_aws/common.yaml#L145), but only when they meet the safety criteria of this RFC. To help with this, a rough audit of "compatibility" can be found in [Appendix A: Compatibility](rfc-128/compatibility.md).
+
+- We can immediately enable automatic deployments for apps marked with a ✅.
+- We will undertake work to address all the issues (⚠) for apps marked with a ❌.
+
+> **We will disable CD for Publishing API**. This was only enabled on a trial basis, and requires non-trivial work to bring it up to the new standard. We should avoid setting a false precedent by leaving CD enabled.
```

Comment:
I'm hoping this will be something everyone on GOV.UK will feel motivated to help with. Some apps may lag behind, especially if they're not worked on and have lots of issues against the criteria.

For me, the really important thing is that we have a practical framework (this RFC) that we can follow to enable automatic deploys for an app. If we have that, we can enable apps at our own pace.

The challenge is how we compromise safety with practicality. I think the balance here about right: many of the apps should only require low effort (~1 day) to enable, so within the realm of possibility.

Short answer: I don't know, I hope we will, but we'll make progress nonetheless.

Does that answer your question?

Reply:
Yup, thank you. Should this be accepted it sounds like it'd be good to have a chat with Senior Tech to see if they're willing to help co-ordinate designating some of the work load across the platform (similar to the exercise we did for dependabot distribution last year but without the lasting commitment)

Reply:
Yes, I agree as well. I have some ideas for an initial rollout:

- Enable all the apps that satisfy the criteria (easy).
- Enable all the apps that are just missing healthchecks (medium).
- Enable all the apps that prevent us deleting the E2E tests (hard).

That leaves a tail of stuff requiring a range of effort to enable. I think that would be a great start, and we can revisit these plans if the RFC is accepted. For now, I'm going to resolve.

cc @bevanloon @richardTowers so you have some awareness.